Mathlib/LinearAlgebra/RootSystem/Finite/Nondegenerate.lean

```diff
@@ -0,0 +1,167 @@
+  zero_lt_apply_root i := P.rootForm_root_self_pos i
+  symm := P.rootForm_symmetric
+  apply_reflection_eq := P.rootForm_reflection_reflection_apply
+
+lemma span_root_finite : Module.Finite R (span R (range P.root)) :=
+  Finite.span_of_finite R <| finite_range P.root
+
+lemma span_coroot_finite : Module.Finite R (span R (range P.coroot)) :=
+  Finite.span_of_finite R <| finite_range P.coroot
+
+lemma span_root_finite_rank : Module.rank R ↥(span R (range ⇑P.root)) < Cardinal.aleph0 := by
+  have := span_root_finite P
+  exact rank_lt_aleph0 R ↥(span R (range ⇑P.root))
+
+lemma finrank_eq_rank_span_root :
+    Module.finrank R (span R (range P.root)) = Module.rank R (span R (range P.root)) := by
+  have := span_root_finite P
+  exact finrank_eq_rank R ↥(span R (range P.root))
+
+lemma finrank_eq_rank_span_coroot :
+    Module.finrank R (span R (range P.coroot)) = Module.rank R (span R (range P.coroot)) := by
+  have := span_coroot_finite P
+  exact finrank_eq_rank R ↥(span R (range P.coroot))
+
+lemma rank_polarization_domRestrict_eq_rank_span_coroot :
+    LinearMap.rank (P.Polarization.domRestrict (span R (range P.root))) =
+      Module.rank R (span R (range P.coroot)) := by
+  refine eq_of_le_of_le (Submodule.rank_mono P.range_polarization_domRestrict_le_span_coroot) ?_
+  have : IsReflexive R N := PerfectPairing.reflexive_right P.toPerfectPairing
+  refine rank_le_of_smul_regular (span R (range P.coroot))
+    (LinearMap.range (P.Polarization.domRestrict (span R (range P.root))))
+    (smul_right_injective N (Ne.symm (ne_of_lt P.prod_rootForm_root_self_pos)))
+    (mem_center_iff.mp (by simp)) fun _ hx => ?_
+  obtain ⟨c, hc⟩ := (mem_span_range_iff_exists_fun R).mp hx
+  rw [← hc, Finset.smul_sum]
+  simp_rw [smul_smul, mul_comm, ← smul_smul]
+  exact Submodule.sum_smul_mem
+    (LinearMap.range (P.Polarization.domRestrict (span R (range P.root)))) c
+    (fun c _ ↦ prod_rootForm_smul_coroot_in_range_domRestrict P c)
+
+lemma rank_coPolarization_domRestrict_eq_rank_span_root :
+    LinearMap.rank (P.CoPolarization.domRestrict (span R (range P.coroot))) =
+      Module.rank R (span R (range P.root)) :=
+  P.flip.rank_polarization_domRestrict_eq_rank_span_coroot
+
+lemma finrank_span_root_le :
+    Module.finrank R (span R (range P.coroot)) ≤ Module.finrank R (span R (range P.root)) := by
+  have h := lift_rank_map_le P.Polarization (span R (range P.root))
+  refine finrank_le_finrank_of_rank_le_rank (le_of_eq_of_le ?_ h) (span_root_finite_rank P)
+  rw [← rank_polarization_domRestrict_eq_rank_span_coroot, ← LinearMap.range_domRestrict]
+
+lemma finrank_span_root_eq :
+    Module.finrank R (span R (range P.coroot)) = Module.finrank R (span R (range P.root)) :=
+  Nat.le_antisymm (finrank_span_root_le P) (by simpa using finrank_span_root_le P.flip)
+
+lemma polarization_domRestrict_kernel_rank_zero : Module.rank R (LinearMap.ker
+    (LinearMap.domRestrict P.Polarization (span R (range P.root)))) = 0 := by
```

Comment:
I usually state everything lattice theoretically when I can because we have such good coverage. So how about instead:
```suggestion
lemma disjoint_rootSpan_ker_polarization :
    Disjoint (span R (range P.root)) (LinearMap.ker P.Polarization) := by
```